controllers/mdutil/util.go

```diff
@@ -425,14 +428,57 @@ func FindNewMachineSet(deployment *clusterv1.MachineDeployment, msList []*cluste
 			return msList[i]
 		}
 	}
-	// new MachineSet does not exist.
+	// New MachineSet does not exist.
 	return nil
 }
 
+// FindNewMachineSetWithRolloutAfter is as FindNewMachineSet but it also accounts for RolloutAfter.
+func FindNewMachineSetWithRolloutAfter(now *metav1.Time, deployment *clusterv1.MachineDeployment, msList []*clusterv1.MachineSet) (*clusterv1.MachineSet, error) {
+	// In rare cases, such as after cluster upgrades, Deployment may end up with
+	// having more than one new MachineSets that have the same template,
+	// see https://github.com/kubernetes/kubernetes/issues/40415
+	// We deterministically choose the oldest new MachineSet with matching template hash.
+	sort.Sort(MachineSetsByCreationTimestamp(msList))
+
+	if deployment.Spec.RolloutAfter != nil {
+		if now.After(deployment.Spec.RolloutAfter.Time) {
+			// TODO (alberto): inlining the format statement within the annotation assigment here
+			// does not set the value. Why?
+			LastRolloutAfter := deployment.Spec.RolloutAfter.Time.UTC().Format(rolloutAfterTimeConversionFormat)
+			deployment.Annotations[clusterv1.LastRolloutAfterAnnotation] = LastRolloutAfter
+		}
+	}
```

Comment:
Why is this part of the code needed? I'm confused on why the annotation is added to the machinedeployment, couldn't we just look and compare at the value of RolloutAfter under Spec?

Reply:
Think of this workflow:

- MachineDeployment is created and Spec.RolloutAfter = nil
  - new MachineSet.Name = name-templateHash 
- Spec.RolloutAfter is set and happened at now-1
  -  old MachineSet.Name = name-templateHash 
  -  new MachineSet.Name = name-templateHash-hash(now-1 )
- Spec.RolloutAfter is set again at now+1

Now you have 2 MachineSets with the same template. Every reconciliation loop until now+1 you want to pick the new one. New one = oldest and created > now-1. Where do you get now-1 from since Spec.RolloutAfter was changed to now+1?

I.e after a rolloutAfter took place and then spec.rolloutAfter is set again to a different value the new MachineSet that needs to be picked every iteration is Oldest and Created > last RolloutAfter

So we need to somehow keep track of the last rolloutAfter. Make sense?